### PR TITLE
helm charts

### PR DESCRIPTION
**Pull request by innerpeacez**
Deploy skywalking to kubernetes using helm charts.

### CODE REVIEW

File: helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-statefulset.yml [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: {{ template "skywalking.fullname" . }}-elasticsearch
+  labels:
```
Comment (hanahmily) [inline helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-statefulset.yml]: add namespace

File: helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-statefulset.yml [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  replicas: {{ .Values.elasticsearch.replicas }}
+  selector:
+    matchLabels:
+      service: {{ .Values.elasticsearch.name }}
+  template:
+    metadata:
+      labels:
+        service: {{ .Values.elasticsearch.name }}
+    spec:
+      terminationGracePeriodSeconds: 300
```
Comment (hanahmily) [inline helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-statefulset.yml]: Parameterize this config.

File: helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-statefulset.yml [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+          # NOTE: the cluster name; update this
+          - name: cluster.name
+            value: elasticsearch-cluster
+          - name: node.name
+            valueFrom:
+              fieldRef:
+                fieldPath: metadata.name
+          # NOTE: This will tell the elasticsearch node where to connect to other nodes to form a cluster
+          - name: discovery.zen.ping.unicast.hosts
+            value: {{ .Values.elasticsearch.name }}:{{ .Values.elasticsearch.ports.tcp }}
```
Comment (hanahmily) [inline helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-statefulset.yml]: Use one value to represent host.

File: helm/charts/6/6.0.0-GA/skywalking/templates/oap/oap-config.yml [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+{{- $oapService := .Values.oap.name -}}
+{{- $namespace := .Release.Namespace -}}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ template "skywalking.fullname" . }}-oap-config
```
Comment (hanahmily) [inline helm/charts/6/6.0.0-GA/skywalking/templates/oap/oap-config.yml]: add namespace

File: helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-statefulset.yml [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+      containers:
+      - name: elasticsearch
+        image: {{ .Values.elasticsearch.image.repository }}:{{ .Values.elasticsearch.image.tag }}
+        ports:
+        - containerPort: {{ .Values.elasticsearch.ports.http }}
+          name: http
+        - containerPort: {{ .Values.elasticsearch.ports.tcp }}
+          name: tcp
+        # NOTE: you can increase this resources
+        resources:
```
Comment (hanahmily) [inline helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-statefulset.yml]: I suggest using `toYaml` to generate resources.

File: helm/charts/6/6.0.0-GA/skywalking/templates/oap/oap-deployment.yml [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ template "skywalking.fullname" . }}-oap
```
Comment (hanahmily) [inline helm/charts/6/6.0.0-GA/skywalking/templates/oap/oap-deployment.yml]: add namespace

File: helm/charts/6/6.0.0-GA/skywalking/templates/oap/oap-deployment.yml [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      serviceAccountName: {{ .Values.oap.name }}
+      containers:
+      - name: {{ .Values.oap.name }}
+        image: {{ .Values.oap.image.repository }}:{{ .Values.oap.image.tag }}
+        ports:
+        - containerPort: {{ .Values.oap.ports.grpc }}
+          name: grpc
+        - containerPort: {{ .Values.oap.ports.rest }}
+          name: rest
+        resources:
```
Comment (hanahmily) [inline helm/charts/6/6.0.0-GA/skywalking/templates/oap/oap-deployment.yml]: `toYaml`

File: helm/charts/6/6.0.0-GA/skywalking/templates/oap/oap-service.yml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ template "skywalking.fullname" . }}-oap
+  labels:
```
Comment (hanahmily) [inline helm/charts/6/6.0.0-GA/skywalking/templates/oap/oap-service.yml]: add namespace

File: helm/charts/6/6.0.0-GA/skywalking/templates/oap/oap-service.yml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    service: {{ .Values.oap.name }}
+spec:
+  ports:
```
Comment (hanahmily) [inline helm/charts/6/6.0.0-GA/skywalking/templates/oap/oap-service.yml]: add `type` to support various types of services

File: helm/charts/6/6.0.0-GA/skywalking/templates/ui/ui-deployment.yml [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ template "skywalking.fullname" . }}-ui
+  labels:
```
Comment (hanahmily) [inline helm/charts/6/6.0.0-GA/skywalking/templates/ui/ui-deployment.yml]: namespace

File: helm/charts/6/6.0.0-GA/skywalking/templates/ui/ui-deployment.yml [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+      labels:
+        app: {{ .Values.ui.name }}
+    spec:
+      containers:
+      - name: {{ .Values.ui.name }}
+        image: {{ .Values.ui.image.repository }}:{{ .Values.ui.image.tag }}
+        ports:
+        - containerPort: {{ .Values.ui.ports.page }}
+          name: page
+        resources:
```
Comment (hanahmily) [inline helm/charts/6/6.0.0-GA/skywalking/templates/ui/ui-deployment.yml]: toYaml

File: helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-pvc.yaml [hunk 12]
```diff
@@ -0,0 +1,27 @@
+{{- if .Values.elasticsearch.persistence.enabled -}}
+{{- if not .Values.elasticsearch.persistence.existingClaim -}}
+apiVersion: v1
+kind: PersistentVolumeClaim
```
[the remark this answers is not in the document]
Comment (innerpeacez) [reply helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-pvc.yaml]: It has been modified， https://github.com/apache/skywalking-kubernetes/pull/8/commits/1c78c801c4133af49e0f7d9d8f856a1a64a86939

File: helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-statefulset.yml (hunk 1 above)
Comment (innerpeacez) [reply helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-statefulset.yml]: The namespace can be specified externally. If you write to death, you will get an error if you want to modify it externally. If you want to get the namespace automatically, you can use the Releace object.
https://github.com/apache/skywalking-kubernetes/pull/8/commits/a00a507ccdad03d7f48bbc9e076101b6e63ba456

**Comment by innerpeacez** [comment]
Hi , About toYaml, If I use the toYaml method, I need to define multiple keys in the values.yaml file, because the deployment and the required keys in the service are different, which causes me to define different keys, but their values are the same, and will cause ui to get the port of oap complicated. 

What should I do ? Tell me ,Thanks.

File: helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-statefulset.yml (hunk 1 above)
Comment (hanahmily) [reply helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-statefulset.yml]: According to my experience, we should add the following piece to specific `namespace`

```
metadata:
  ...
  namespace: {{ .Release.Namespace }}
```
Is this a mandatory configuration for deploying charts into a non-default namespace, or not?

**Comment by hanahmily** [comment]
>  the deployment and the required keys in the service are different,

My suggestion is "toYaml" resource(cpu, memory) configuration instead of all of the yamls.

File: helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-statefulset.yml (hunk 1 above)
Comment (innerpeacez) [reply helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-statefulset.yml]: I understand what you said, but when we deploy skywalking, we may not deploy to namespace=skywaling or other specific namespaces. We may deploy to dev, test or prod environments. 

If we write the namespace configuration, we deploy to different environments, you need to modify the charts every time. I suggest using the command line to specify the namespace. 

` hellm install skywaling --namespace=xxx`

However, if the namespace is difined in the charts, the command line will return an error.

Comment (innerpeacez) [reply helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-statefulset.yml]: {{ .Release.Namespace }} can get input from the command line

Comment (innerpeacez) [reply helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-statefulset.yml]: not mandatory, kubernetes will automatically populate the namespace configuration.{{ .Release.Namespace }}  also can get this.

**Comment by innerpeacez** [comment]
Except for namespace, other Changes requested have been modified.

About namespace ,I have two ways.

- now charts
- add the namespace configuration to each file and maintain this configuration with values.yaml
`
helm install skywalking --set=""
`
I recommend using the first method, I refer to helm/charts repo.

Help review again , thanks

File: helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-statefulset.yml (hunk 1 above)
Comment (hanahmily) [reply helm/charts/6/6.0.0-GA/skywalking/templates/elasticsearch/elasticsearch-statefulset.yml]: > Except for namespace, other Changes requested have been modified.
> 
> About namespace ,I have two ways.
> 
> * now charts
> * add the namespace configuration to each file and maintain this configuration with values.yaml
>   `helm install skywalking --set=""`
>   I recommend using the first method, I refer to helm/charts repo.
> 
> Help review again , thanks

You are right. Most of the charts in helm repo don't contain `namespace` in their artifacts. I used to set it but it seems not the best practice.

**Pull request merged**